src/components/PopOver/index.js

```diff
@@ -0,0 +1,240 @@
+    const topPosition = elTop - SPACE_FROM_MOUSE - height;
+
+    const spaceFromEdge =
+      windowWidth > MOBILE_MAX_WIDTH
+        ? MIN_SPACE_FROM_EDGE
+        : MOBILE_MIN_SPACE_FROM_EDGE;
+
+    const containerTop = offsetTop + spaceFromEdge;
+    const containerBottom = offsetTop + clientHeight - spaceFromEdge;
+    const containerLeft = offsetLeft + spaceFromEdge;
+    const containerRight = offsetLeft + clientWidth - spaceFromEdge - width;
+
+    const topPisitionWithFallback =
+      topPosition - spaceFromEdge >= Math.max(viewportTop, containerTop)
+        ? topPosition
+        : bottomPosition;
+
+    return {
+      x: Math.min(
+        Math.max(0 + spaceFromEdge, mouseX - width / 2, containerLeft),
+        windowWidth - spaceFromEdge - width,
+        containerRight
+      ),
+      y:
+        bottomPosition + height + spaceFromEdge <=
+        Math.min(viewportBottom, containerBottom)
+          ? bottomPosition
+          : topPisitionWithFallback
+    };
+  }
+
+  static getDerivedStateFromProps(props) {
+    if (!props.isVisible) {
+      return {
+        width: 0,
+        height: 0
+      };
+    }
+
+    return null;
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      width: 0,
+      height: 0,
+      windowScroll: 0,
+      windowWidth: 0,
+      windowHeight: 0
+    };
+
+    this.myRef = React.createRef();
+  }
+
+  componentDidUpdate() {
+    if (this.myRef.current) {
+      this.setDimensions();
+    }
+  }
+
+  setDimensions = () => {
+    const { isVisible } = this.props;
+    const {
+      windowScroll,
+      windowWidth,
+      windowHeight,
+      width,
+      height
+    } = this.state;
+    const dimensions = {};
+    if (global.window && isVisible) {
+      const {
+        scrollTop,
+        clientWidth,
+        clientHeight
+      } = global.window.document.documentElement;
+
+      if (scrollTop !== windowScroll) {
+        dimensions.windowScroll = scrollTop;
+      }
+
+      if (clientWidth !== windowWidth) {
+        dimensions.windowWidth = clientWidth;
+      }
+
+      if (clientHeight !== windowHeight) {
+        dimensions.windowHeight = clientHeight;
+      }
+    }
+
+    if (this.myRef.current) {
+      const { clientWidth, clientHeight } = this.myRef.current;
+
+      if (width !== clientWidth && clientWidth) {
+        dimensions.width = clientWidth;
+      }
+
+      if (height !== clientHeight && clientHeight) {
+        dimensions.height = clientHeight;
+      }
+    }
+
+    if (
+      Object.keys(dimensions).length &&
+      dimensions.width &&
+      dimensions.height
+    ) {
+      this.setState(dimensions);
+      return true;
+    }
+
+    return false;
+  };
+
+  render() {
+    const {
+      children,
+      isVisible,
+      position,
+      reduceTop,
+      reduceBottom
+    } = this.props;
+    const pos =
+      isVisible && !this.setDimensions()
+        ? PopOver.calculatePosition(position, this.state, {
```

Comment:
It seems pretty heavy to potentially call both `this.setDimensions` and `PopOver.calculatePosition` each time this component renders... I want to better understand why this is required, rather than relying on the component which is associated with the `PopOver` ("being popped over") to determine its positioning.

Reply:
If setDimensions returns false we don't call calculatePosition and when isVisible is false we don't call any of the two functions in order to make it lighter.
setDimensions checks if there is a change in the tooltip size view port size and position which are crucial for determining the correct position of the PopOver.
Does this answer your question?

Reply:
Yes, thank you